File: 백준/Gold/9935. 문자열 폭발/문자열 폭발.py

a1 = input()
a2 = list(input())

stack = []

for i in range(len(a1)):
    stack.append(a1[i]) # 스택에 하나씩 추가
    if stack[-len(a2):] == a2: # 스택의 마지막이 m 문자열과 같으면
        del stack[-len(a2):] # 삭제

if stack: print("".join(stack))
else: print("FRULA")

File: 백준/Gold/9935. 문자열 폭발/문자열 폭발.java
import java.util.Arrays;
import java.util.Scanner;
import java.util.Stack;

public class Main {

  static String str;
  static String compare;
  static Stack<Character> stack = new Stack<>();
  public static void main(String[] args) {
    Scanner sc = new Scanner(System.in);
    str = sc.nextLine();
    compare = sc.nextLine();
    int m = compare.length();

    for(int i = 0; i < str.length(); i++){
      stack.push(str.charAt(i));
      if(stack.size()< m){
        continue;
      }
      boolean flag = true;
      // 1 2 3      4123
      for(int j = 0; j < m; j++){
        if(compare.charAt(j) != stack.get(stack.size()-m+j)){
          flag = false;
          break;
        }
      }
      if (flag){
        for (int j = 0; j < m; j++){
          stack.pop();
        }
      }

    }
    if (stack.isEmpty()) {
      System.out.println("FRULA");
    } else {
      StringBuilder sb = new StringBuilder();
      for (int i = 0; i < stack.size(); i++) {
        sb.append(stack.get(i));
      }
      System.out.println(sb.toString());
    }

  }
}
